### PR TITLE
Add support for opening folders.

### LINKED ISSUE #941: File Open Dialog Should Support Opening Folders
### Is your feature request related to a problem? Please describe:
It is only possible for the user to open a file, it is not possible for the user to open a folder.

### Is it possible to construct a solution with the existing API?
No

### Describe the solution you'd like to see:
It should be possible to show a dialog for the user to select a folder, either through the existing ShowOpenDialog, or a new ShowOpenFolderDialog.

### PR DESCRIPTION
**Pull request by andydotxyz**
New constructor and show helper for folders.
This re-uses the file dialog using a standard filter for folder type (largely an internal detail for signalling).
For desktop this sets the current directory as the "selected" item so "Open" just picks this.
This does open mobile pickers as well, though the lack of ListableURI in mobile means it does not return correctly.

Proposing a new PR for completing ListableURI for mobile so that it does not hold back this functionality on desktop.

Fixes #941

### Checklist:

- [x] Tests included.
- [x] Lint and formatter run with no errors.
- [x] Tests all pass.
- [x] Public APIs match existing style.

### CODE REVIEW

File: cmd/fyne/internal/mobile/dex.go [hunk 1]
```diff
@@ -6,179 +6,183 @@
...
-	`AQACAQEEkCABkCA0gIAEsBY1AdAWAAEBAQaQIDaAgATQGDcB7BgAAQEBB5AgOICABJQZOQ` +
-	`GwGQABAQMIkCA6gIAEzBk7AegZAQH8GQEBkBoAAQEBCZAgPoCABMAbPwHcGwYCEQcKGgEa` +
-	`ARoBGgEaAQoQAgECQIGABIgdAYggrB0BiCDEHQGIINwdAYgg9B0BiCCMHgGIIKQeBYICAA` +
-	`UIvB4ECJAfAYICAAGCAgABggIAAQKsHwQCsCABCNQgAQjwIEgAjCEBAOQhAQDoIggAjCMH` +
-	`BLAjAQH0IwYAwCQRAAAAAAAAAAEAAAAAAAAAAQAAALMAAABwAAAAAgAAAC0AAAA8AwAAAw` +
-	`AAADkAAADwAwAABAAAABIAAACcBgAABQAAAGEAAAAsBwAABgAAAAYAAAA0CgAAAxAAAAUA` +
-	`AAD0CgAAASAAACAAAAAwCwAABiAAAAUAAAAcEwAAARAAAB8AAABsEwAAAiAAALMAAACIFA` +
-	`AAAyAAACAAAACGIAAABCAAAAYAAADTIQAABSAAAAEAAAD/IQAAACAAAAYAAAAMIgAAABAA` +
-	`AAEAAADoIgAA` +
+var dexStr = `ZGV4CjAzNQAClnuWucRCmShMpICNQdPrq7aZEIeogCKMJAAAcAAAAHhWNBIAAAAAAAAAAL` +
```
Comment (Jacalz) [inline cmd/fyne/internal/mobile/dex.go:9]: What is this doing and why is it changed? Unfortunately just seems like a bunch of scramble characters to an untrained eye :sweat_smile: A comment might be a good addition :)

**Review by Jacalz — COMMENTED** [review]
Just one small comment, otherwise it looks good to me. It is great to see this finally landing :+1:

File: cmd/fyne/internal/mobile/dex.go (hunk 1 above)
Comment (andydotxyz) [reply cmd/fyne/internal/mobile/dex.go:9]: The comment is at the top of the file, it is generated by gendex.
Every time we change the Java bootstrap code in the mobile dep we have to re-package the compiled object code into this "dex" file (android talk for java bytecode).

Comment (Jacalz) [reply cmd/fyne/internal/mobile/dex.go:9]: Sorry. I thought I was looking at the top of the file when I saw the `package main`.

**Review by Jacalz — APPROVED** [review]
Looks great. Thank you :)

File: cmd/fyne_demo/screens/window.go [hunk 2]
```diff
@@ -172,6 +172,18 @@ func loadDialogGroup(win fyne.Window) *widget.Card {
...
 		}),
+		widget.NewButton("Folder Open", func() {
+			dialog.ShowFolderOpen(func(list fyne.ListableURI, err error) {
+				if list == nil {
+					return
+				}
+				if err != nil {
+					dialog.ShowError(err, win)
+					return
+				}
```
Comment (stuartmscott) [inline cmd/fyne_demo/screens/window.go:183]: This ordering means the error message will not get shown

**Review by stuartmscott — CHANGES_REQUESTED** [review]

File: cmd/fyne_demo/screens/window.go (hunk 2 above)
Comment (andydotxyz) [reply cmd/fyne_demo/screens/window.go]: Thanks, fixed

**Review by Jacalz — APPROVED** [review]

**Review by stuartmscott — APPROVED** [review]

**Pull request merged**